tests/net/socket/tcp/src/main.c

```diff
@@ -723,6 +723,108 @@ void test_open_close_immediately(void)
 	k_sleep(TCP_TEARDOWN_TIMEOUT);
 }
 
+#define TCP_CLOSE_FAILURE_TIMEOUT 60000
+
+void test_close_obstructed(void)
+{
+	/* Test if socket closing even when there is not communication
+	 * possible any more
+	 */
+	uint32_t start_time, time_diff;
+	int count_before = 0, count_after = 0;
+	struct sockaddr_in c_saddr;
+	struct sockaddr_in s_saddr;
+	struct sockaddr addr;
+	socklen_t addrlen = sizeof(addr);
+	int c_sock;
+	int s_sock;
+	int new_sock;
+	int dropped_packets_before = 0;
+	int dropped_packets_after = 0;
+
+	prepare_sock_tcp_v4(CONFIG_NET_CONFIG_MY_IPV4_ADDR, ANY_PORT,
+			    &c_sock, &c_saddr);
+	prepare_sock_tcp_v4(CONFIG_NET_CONFIG_MY_IPV4_ADDR, SERVER_PORT,
+			    &s_sock, &s_saddr);
+
+	test_bind(s_sock, (struct sockaddr *)&s_saddr, sizeof(s_saddr));
+	test_listen(s_sock);
+
+	zassert_equal(connect(c_sock, (struct sockaddr *)&s_saddr,
+				sizeof(s_saddr)),
+				0, "connect not succeed");
+	test_accept(s_sock, &new_sock, &addr, &addrlen);
+
+	/* We should have two contexts open now */
+	net_context_foreach(calc_net_context, &count_before);
+
+	/* Break the communication */
+	loopback_set_packet_drop_ratio(1.0f);
```

Comment:
If we want to verify if FIN is actually retransmitted, I suggest a simpler test:
```
	/* Break the communication */
	loopback_set_packet_drop_ratio(1.0f);

	test_close(c_sock);
   
	k_sleep(K_MSEC(200));

	/* Server socket should not notify connection closed at this point yet */
	ret = recv(new_sock, &buf, sizeof(buf), MSG_DONTWAIT);
	zassert_equal(ret, -1, "not -1");
	zassert_equal(errno, EAGAIN, "not EAGAIN");

	loopback_set_packet_drop_ratio(0.0f);
	/* Give the retransmission timer a chance to kick in */
	k_sleep(K_MSEC(500));

	/* Server socket should report closed connection now */
	ret = recv(new_sock, &buf, sizeof(buf), MSG_DONTWAIT);
	zassert_equal(ret, 0, "not 0");

	test_close(new_sock);
	test_close(s_sock);
```

Reply:
Thanks for the suggestion. Looking at this, we might want to make 2 different tests. The existing test to validate the proper closure of the context, the other to count check the FIN retry as you suggested. Do you agree?

Reply:
net context close is already verified by `test_open_close_immediately`, I don't think if there's a need to duplicate that.

Reply:
I think it is better to have both the context closing and counting of FIN packets checked